// @cparra/apexdocs/lib/model/ApexModel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Settings_1 = require("../Settings");
class ApexModel {
    constructor() {
        this.nameLine = '';
        this.description = '';
        this.author = '';
        this.date = '';
        this.returns = '';
        this.example = '';
        this.scope = '';
    }
    getNameLine() {
        return this.nameLine;
    }
    getInameLine() {
        return this.nameLineIndex;
    }
    setNameLine(nameLine, iLine) {
        this.nameLine = nameLine.trim();
        this.nameLineIndex = iLine;
        this.parseScope();
    }
    getDescription() {
        return this.description == null ? '' : this.description;
    }
    setDescription(description) {
        this.description = description;
    }
    getAuthor() {
        return this.author == null ? '' : this.author;
    }
    setAuthor(author) {
        this.author = author;
    }
    getDate() {
        return this.date == null ? '' : this.date;
    }
    setDate(date) {
        this.date = date;
    }
    getReturns() {
        return this.returns == null ? '' : this.returns;
    }
    setReturns(returns) {
        this.returns = returns;
    }
    getExample() {
        return this.example == null ? '' : this.example;
    }
    setExample(example) {
        this.example = example;
    }
    getScope() {
        return this.scope == null ? '' : this.scope;
    }
    setScope(scope) {
        this.scope = scope;
    }
    parseScope() {
        this.scope = '';
        const str = this.getScopeFromSettings(this.nameLine);
        if (str != null) {
            this.scope = str;
        }
    }
    getScopeFromSettings(str) {
        str = str.toLowerCase();
        for (const currentScope of Settings_1.default.getInstance().getScope()) {
            if (str.toLowerCase().includes(currentScope.toLowerCase() + ' ')) {
                return currentScope;
            }
        }
        return null;
    }
}
exports.default = ApexModel;
