// @cparra/apexdocs/lib/Settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Settings {
    constructor() {
        this.desiredScope = ['global', 'public'];
        this.outputDir = 'docs';
        this.configPath = null;
        this.shouldGroup = true;
        this.processor = null;
    }
    static getInstance() {
        if (!Settings.instance) {
            Settings.instance = new Settings();
        }
        return Settings.instance;
    }
    setScope(desiredScope) {
        this.desiredScope = desiredScope;
    }
    getScope() {
        return this.desiredScope;
    }
    setOutputDir(outputDir) {
        this.outputDir = outputDir;
    }
    getOutputDir() {
        return this.outputDir;
    }
    setDocsProcessor(processor) {
        this.processor = processor;
    }
    getDocsProcessor() {
        return this.processor;
    }
    setConfigPath(configPath) {
        this.configPath = configPath;
    }
    getConfigPath() {
        return this.configPath;
    }
    setShouldGroup(shouldGroup) {
        this.shouldGroup = shouldGroup;
    }
    getShouldGroup() {
        return this.shouldGroup;
    }
}
exports.default = Settings;
